### DISCUSSION TITLE
Jakarta-REST-4.0-Discussion is not publicly accessible

### DISCUSSION THREAD

**Discussion by anbusampath** [eclipse-ee4j/jersey#5530] [General]
https://github.com/eclipse-ee4j/jaxrs-api/wiki/Jakarta-REST-4.0-Discussion - this page not accessible. Link is provided in roadmap page at the bottom - https://github.com/eclipse-ee4j/jersey/wiki/Road-Map

**Comment by jansupol** [comment]
It used to be there, but it is not any longer. Perhaps it is this project now https://github.com/jakartaee/rest/projects/2

**Reply by jansupol** [reply]
Or this https://github.com/jakartaee/rest/wiki/Jakarta-REST-4.0-Discussion-(EE-11)

**Reply by anbusampath** [reply]
just watched Ivar's Jfokus [talk](https://www.youtube.com/watch?v=mRRZ76ZBV_M) and also rest-dev mailing [list](https://www.eclipse.org/lists/rest-dev/msg00206.html), it's looks like REST 4 not happening in Jakarta EE 11 timeline.

**Reply by jansupol** [reply]
The current plan is to publish it as 3.2, and keep deprecated `@Context` support to stay backward compatible. But there is a slight possibility it won't happen in time, let's see.

**Reply by jansupol** [reply]
So it's going to be 4.0, anyway

**Reply by anbusampath** [reply]
Great, is there place to refer list of features tagged for 4.0 release?

**Reply by anbusampath** [reply]
I belive below links are slightly outdated.
https://projects.eclipse.org/projects/ee4j.rest/releases/4.0.0
https://jakarta.ee/specifications/restful-ws/4.0/

**Reply by jansupol** [reply]
This is what is going to be in 4.0: https://github.com/jakartaee/specifications/pull/705

**Reply by martinofcourse** [reply]
I am sorry for piggybacking on this discussion, but couldn't find a better place to post.

If `@Context` is going to be deprecated, then how would one accomplish what `@Context Application application` does in an implementation-neutral way? For instance, when using `SeBootstrap`, the `Application` subclass is instantiated using `new` which thwarts CDI and `@Inject` won't work.

```Java
static CompletionStage<Instance> start(final Class<? extends Application> clazz, final Configuration configuration)
```

The comments to the above call explicitly state that `Injection is not supported`.

This also seem to effectively deprive us of the ability to perform eager initialization without observing events and writing an extension.

**Reply by martinofcourse** [reply]
Also, in `jakarta.ws.rs.SeBootstrap` I see

```Java
        default void stopOnShutdown(Consumer<StopResult> consumer) {
            Runtime.getRuntime().addShutdownHook(new Thread(() -> stop().thenAccept(consumer)));
        }
 ```

I could never get the `consumer` in the above to run before the JVM shuts down. I think if we want the `CompletionStage<StopResult> stop();` and the `consumer` code to run **before** the JVM shuts down, we need something like this instead:

```Java
        default void stopOnShutdown(Consumer<StopResult> consumer) {
            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                stop().toCompletableFuture().thenAccept(consumer).join();
            }));
        }
 ```
 
Also, the `CompletionStage` twist on the `stop()` method seems superfluous when we are running off the new thread in the shutdown hook.
